frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx

```diff
@@ -0,0 +1,428 @@
+    isDefaultReceiver,
+    receiverToEdit,
+    route.routes,
+  );
+
+  const { formValues, dispatchFormChange } = useFormStore(INITIAL_STATE);
+  const SubForm = subformFactory(formValues.receiverType);
+
+  const isFormInvalid =
+    _.isEmpty(formValues.receiverName) ||
+    _.isEmpty(formValues.receiverType) ||
+    SubForm.isFormInvalid(formValues);
+
+  const save = (e) => {
+    e.preventDefault();
+
+    // Update Receivers
+    const newReceiver = createReceiver(formValues, SubForm.createReceiverConfig, receiverToEdit);
+    _.update(config, 'receivers', (receivers = []) => {
+      if (editReceiverNamed) {
+        const index = _.findIndex(receivers, { name: editReceiverNamed });
+        receivers.splice(index, 1, newReceiver);
+      } else {
+        receivers.push(newReceiver);
+      }
+      return receivers;
+    });
+
+    // Update Route & RouteLabels
+    if (isDefaultReceiver) {
+      _.set(route, 'receiver', newReceiver.name);
+    }
+
+    const newRoute = _.isEmpty(formValues.routeLabels)
+      ? undefined
+      : createRoute(newReceiver, formValues.routeLabels);
+    _.update(route, 'routes', (routes = []) => {
+      if (editReceiverNamed) {
+        const index = _.findIndex(routes, { receiver: editReceiverNamed });
+        if (index !== -1) {
+          if (!newRoute) {
+            // no routing labels for receiver, remove old route
+            routes.splice(index, 1);
+          } else {
+            // update receiver's route with new route/labels
+            routes.splice(index, 1, newRoute);
+          }
+        } else if (newRoute) {
+          // receiver didn't have a prior route, so add new route
+          routes.push(newRoute);
+        }
+      } else if (newRoute) {
+        // add route for new receiver
+        routes.push(newRoute);
+      }
+      return routes;
+    });
+
+    // Update 'alertmanager-main' Secret with new alertmanager.yaml configuration
+    setInProgress(true);
+    patchAlertmanagerConfig(secret, config).then(
+      () => {
+        setErrorMsg('');
+        setInProgress(false);
+        history.push('/monitoring/alertmanagerconfig');
+      },
+      (err) => {
+        setErrorMsg(err.message);
+        setInProgress(false);
+      },
+    );
+  };
+
+  return (
+    <div className="co-m-pane__body co-m-pane__form">
+      <Helmet>
+        <title>{titleVerb} Receiver</title>
+      </Helmet>
+      <form className="co-m-pane__body-group" onSubmit={save}>
+        <h1 className="co-m-pane__heading">
+          {titleVerb} {isDefaultReceiver && 'Default'} Receiver
+        </h1>
+        <div className="form-group">
+          <label className="control-label co-required">Receiver Name</label>
+          <input
+            className="pf-c-form-control"
+            type="text"
+            value={formValues.receiverName}
+            onChange={(e) =>
+              dispatchFormChange({
+                type: 'handleChangeEvent',
+                payload: { ...e },
```

Comment:
```suggestion
                payload: e,
```

Reply:
Hi @spadgett, when I pass just the `e` by reference, for some reason every other 'change' results in `action.payload.target` being `null`.  So entering a receiver name 'd' works, 2nd char 'e' target is `null`, third 'f' works, forth `null`, etc...
Only passing in a new object seems to solve this issue.

Reply:
Honestly, I think it would be better to make these:

```ts
dispatchFormChange({
  type: 'setFormValues',
  payload: {
    receiverName: e.target.value,
  },
});
```

and remove `handleChangeEvent`. The dispatch function really shouldn't know anything about the form events, imo. Only the code here should know about the input element.

Reply:
Ok, done.  Also removed unneeded `useFormStore` wrapper method and used ` React.useReducer` directly.